README.md
# MyAnimeListUserListsMerger
Creates a CSV that is a merged anime list of multiple MAL users. In this file there should be the name of a MAL user or the link to their profile in each line. It will give out a "MergedLists.txt", that you can import to as example Excel or Calc with "^" as the cell seperator.

A little tutorial on how I use this program:
First I go to overmes' MAL Affinity Search: http://affinity.animesos.net/
There I start multiple searches with differencing shared titles amounts. I suggest starting with about 10% of your completed anime and then increase until the lowest affinity with users you get is 75%.
Then I use the Firefox addon "Link Gopher" to extract all the profile links on the page (for some reason I have to reload the website after the searches).
After that I start the program, paste them into the users text area, enter my MAL username and press start.
Now I wait until it's finished, import the txt file into Excel, use ^ as cell seperator and also set that it's a UTF-8 file.
Optionally I first sort after score count and move every anime with a score count less than 3 to an extra table.
Then I sort them after weighted score.

Old Bitbucket Repo: https://bitbucket.org/meesterP_/myanimelistuserlistsmerger/src/master/


Changelog.md
2.0.0:

A GUI

Better encapsulated and improved code

Ability to ignore the completed anime of multiple users

1.2.4:

Code improvements


XMLParser.java
package xmlparser;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.concurrent.TimeUnit;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.apache.commons.io.FileUtils;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

public class XMLParser {

    private int userCounter = 0;
    private boolean enteredUserToBeIgnored = false;
    private ArrayList<Integer> arrayListOfIgnoredEntries;
    private int sleepTime = 200;

    public static void main(String[] args) {
        XMLParser myInstance = new XMLParser();
        try {
            myInstance.createIgnoredAnimeList();
        } catch (IOException | SAXException | ParserConfigurationException e) {
            throw new RuntimeException(e);
        }
    }

    private void createIgnoredAnimeList() throws IOException, ParserConfigurationException, SAXException {
        Scanner sc = new Scanner(System.in);
        System.out.println("If you want the completed entries of a user to be excluded enter their name or don't and then press enter:");
        String ignoreCompletedUser = sc.nextLine();
        System.out.println("Enter amount of waiting in miliseconds between each user. Standard is 200. Increase if you are getting 429 errors (500ms should be enough), decrease if you think you are lucky.");
        sleepTime = sc.nextInt();
        if (sleepTime < 0) sleepTime = 200;
        sc.close();
        if (ignoreCompletedUser.length() > 0) {
            enteredUserToBeIgnored = true;
            String myAnimeListUserURL = "https://myanimelist.net/malappinfo.php?u=" + ignoreCompletedUser + "&status=all&type=anime";
            File xmlFile = new File("animelist", "tmp");
            URL url = new URL(myAnimeListUserURL);

            System.out.println("Processing entered User: " + ignoreCompletedUser);
            FileUtils.copyURLToFile(url, xmlFile);

            // Create the hashmap for this user
            ArrayList<Integer> animeList = new ArrayList<Integer>();

            // Parse the xml
            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(xmlFile);
            doc.getDocumentElement().normalize();
            NodeList nList = doc.getElementsByTagName("anime");

            // go through the parsed list and add it to the users list
            for (int i = 0; i < nList.getLength(); i++) {
                Node nNode = nList.item(i);

                // if its a valid anime node, create a row, extract the data and save it into the user
                // animelist
                if (nNode.getNodeType() == Node.ELEMENT_NODE) {
                    Element elem = (Element) nNode;
                    // only add anime if it is completed
                    if (Integer.parseInt(elem.getElementsByTagName("my_status").item(0).getTextContent()) == 2) {
                        Integer key = Integer.parseInt(elem.getElementsByTagName("series_animedb_id").item(0).getTextContent());
                        // put the anime into the row
                        animeList.add(key);
                        arrayListOfIgnoredEntries = animeList;
                    }
                }
            }
        }
        parseXML();
    }

    private void parseXML() throws IOException, ParserConfigurationException, SAXException {
        String readLine;
        String newReadLine = "";
        BufferedReader reader = new BufferedReader(new FileReader("malusers.txt"));
        ArrayList<HashMap<Integer, Row>> listOfMaps = new ArrayList<>();
        while ((readLine = reader.readLine()) != null) {
            int lastSlashPosition = readLine.lastIndexOf("/");
            if (lastSlashPosition != -1) {
                for (int i = lastSlashPosition + 1; i < readLine.length(); i++) {
                    newReadLine += readLine.charAt(i);
                }
                readLine = newReadLine;
                newReadLine = "";
            }
            HashMap<Integer, Row> aMap = getAndParseXmlForUser(readLine);
            listOfMaps.add(aMap);
        }

        HashMap<Integer, Row> combinedMap = listOfMaps.stream().reduce(new HashMap<>(), this::mergeMaps);

        reader.close();
        removeEntriesOfEnteredUser(arrayListOfIgnoredEntries, combinedMap);
    }

    private HashMap<Integer, Row> getAndParseXmlForUser(String user) throws IOException, ParserConfigurationException, SAXException {
        String myAnimeListUserURL = "https://myanimelist.net/malappinfo.php?u=" + user + "&status=all&type=anime";
        File xmlFile = new File("animelist", "tmp");
        URL url = new URL(myAnimeListUserURL);

        // sleep is necessary because the MAL API complains at too many requests
        // skip for first user
        try {
            TimeUnit.MILLISECONDS.sleep(sleepTime);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        userCounter++;
        System.out.println("Processing User #" + userCounter + ": " + user);
        FileUtils.copyURLToFile(url, xmlFile);

        // Create the hashmap for this user
        HashMap<Integer, Row> animeMap = new HashMap<>();

        // Parse the xml
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        Document doc = dBuilder.parse(xmlFile);
        doc.getDocumentElement().normalize();
        NodeList nList = doc.getElementsByTagName("anime");

        // go through the parsed list and add it to the users list
        for (int i = 0; i < nList.getLength(); i++) {
            Node nNode = nList.item(i);

            // if its a valid anime node, create a row, extract the data and save it into the user
            // animelist
            if (nNode.getNodeType() == Node.ELEMENT_NODE) {
                Element elem = (Element) nNode;
                // only add anime if it is not PTW
                if (Integer.parseInt(elem.getElementsByTagName("my_status").item(0).getTextContent()) != 6) {
                    Row row = new Row();
                    row.name = elem.getElementsByTagName("series_title").item(0).getTextContent();
                    row.count = 1;
                    // only add score and scorecount if score is not 0
                    if (Integer.parseInt(elem.getElementsByTagName("my_score").item(0).getTextContent()) == 0) {
                        row.score = 0;
                        row.scoreCount = 0;
                    } else {
                        row.score = Double.parseDouble(elem.getElementsByTagName("my_score").item(0).getTextContent());
                        row.scoreCount = 1;
                    }

                    Integer key = Integer.parseInt(elem.getElementsByTagName("series_animedb_id").item(0).getTextContent());
                    // put the anime into the row
                    animeMap.put(key, row);
                }
            }
        }
        return animeMap;
    }

    private HashMap<Integer, Row> mergeMaps(HashMap<Integer, Row> mapA, HashMap<Integer, Row> mapB) {
        HashMap<Integer, Row> newMap = new HashMap<>();
        newMap.putAll(mapA);

        for (Map.Entry<Integer, Row> entry : mapB.entrySet()) {
            if (newMap.containsKey(entry.getKey())) { // If the key is already in the map, we have a duplicate!

                // We need to update the existing entry

                // Only proceed if the user gave a rating
                if (entry.getValue().score != 0) {
                    if (newMap.get(entry.getKey()).score > 0) {
                        newMap.get(entry.getKey()).score = ((newMap.get(entry.getKey()).score * newMap.get(entry.getKey()).scoreCount) + entry.getValue().score) / (newMap.get(entry.getKey()).scoreCount + 1);
                        newMap.get(entry.getKey()).scoreCount += 1;
                    } else {
                        newMap.get(entry.getKey()).score = entry.getValue().score;
                        newMap.get(entry.getKey()).scoreCount = 1;
                    }
                }
                newMap.get(entry.getKey()).count += 1;
                // this should be it
            } else { // Key of map 2 is not in the new map, just add the non duplicate
                newMap.put(entry.getKey(), entry.getValue());
            }
        }
        return newMap;
    }

    private void removeEntriesOfEnteredUser(ArrayList<Integer> arrayListOfIgnoredEntries, HashMap<Integer, Row> combinedMap) {
        if (enteredUserToBeIgnored) {
            HashMap<Integer, Row> newMap = new HashMap<>();
            newMap.putAll(combinedMap);

            System.out.println("Removing entries");
            for (Integer elem : arrayListOfIgnoredEntries) {
                // if we find the same key in both maps, remove the entry
                if (newMap.containsKey(elem)) {
                    newMap.remove(elem);
                }
            }
            writeCSV(newMap);
        } else {
            writeCSV(combinedMap);
        }
}

    private void writeCSV(HashMap<Integer, Row> map) {
        double factorizedScore;
        double scoreCountNormalization;
        double weightedScore;
        final double FACTOR = 0.866;
        String link;
        int maxCount = 1;
        int titleAmountWithScore = 0;
        double totalScore = 0;

        for (Map.Entry<Integer, Row> entry : map.entrySet()) {
            if (entry.getValue().count > maxCount) {
                maxCount = entry.getValue().count;
            }
            if (entry.getValue().score > 0) {
                titleAmountWithScore++;
            }
            totalScore += entry.getValue().score;
        }

        System.out.println("Writing File");
        String result = "Name^Score^Count^Score Count^Weighted Score^Link" + "\n";

        for (Map.Entry<Integer, Row> entry : map.entrySet()) {
            link = "https://myanimelist.net/anime/" + entry.getKey();
            factorizedScore = (double) entry.getValue().score * FACTOR;
            scoreCountNormalization = (double) entry.getValue().count / maxCount;
            weightedScore = factorizedScore + scoreCountNormalization * (1.0 - FACTOR) * 10.0;
            result = result + entry.getValue() + "^" + weightedScore + "^" + link + "\n";
        }

        System.out.println("Max count: " + maxCount);
        System.out.println("Average rating: " + totalScore / titleAmountWithScore);

        try {
            FileUtils.writeStringToFile(new File("MergedLists.txt"), result, "UTF-8");
            System.out.println("Finished!");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}


src/main/java/gui/Main.java
package gui;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import merger.Merger;
import org.xml.sax.SAXException;

import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Main extends Application {
    private TextArea usersTextArea;
    private TextArea ignoredUsersTextArea;
    private TextField sleepTimeTextField;
    private Label averageScoreLabel;
    private Label maxCountLabel;
    private ProgressBar progressBar;
    private Label progressLabel;

    public static void main(String[] args) {
        launch(args);
    }

    @Override
    public void start(Stage stage) {
        //Scene and Stage Set-Up
        stage.setTitle("MyAnimeList User Lists Merger");
        final VBox root = new VBox();
        Scene scene = new Scene(root);

        //Creation of Scene Objects
        Label ignoredUserLabel = new Label("Ignored Users:");
        ignoredUsersTextArea = new TextArea();

        Label sleepTimeLabel = new Label("Sleep Time (ms):");
        sleepTimeTextField = new TextField("200");

        Button startButton = new Button("Start");
        startButton.setOnAction((ActionEvent e) -> startButtonClick());

        Label usersLabel = new Label("Users:");
        usersTextArea = new TextArea();

        averageScoreLabel = new Label("Average Score: ");

        maxCountLabel = new Label("Max Count: ");

        progressBar = new ProgressBar();
        progressBar.prefWidthProperty().bind(root.widthProperty());

        progressLabel = new Label("Press button to start the process!");

        //Placing Scene Objects in Scene
        root.getChildren().add(usersLabel);
        root.getChildren().add(usersTextArea);
        root.getChildren().add(ignoredUserLabel);
        root.getChildren().add(ignoredUsersTextArea);
        root.getChildren().add(sleepTimeLabel);
        root.getChildren().add(sleepTimeTextField);
        root.getChildren().add(averageScoreLabel);
        root.getChildren().add(maxCountLabel);
        root.getChildren().add(progressBar);
        root.getChildren().add(progressLabel);
        root.getChildren().add(startButton);


        //Starting the stage
        stage.setScene(scene);
        stage.show();
    }

    private void startButtonClick() {
        progressBar.setProgress(0.0);
        progressLabel.setText("Progress: 0%");
        new Thread(() -> {
            Merger pap = new Merger(Integer.parseInt(sleepTimeTextField.getText()), usersTextArea.getText(), ignoredUsersTextArea.getText());
            int userAmount = pap.getUsers().size();
            int counter = 1;

            //Create Map with anime of all the users
            for (String user : pap.getUsers()) {
                try {
                    pap.addMapToMergedAnimeListsMap(pap.createMapOfAnime(user));
                } catch (IOException | ParserConfigurationException | SAXException | InterruptedException e) {
                    e.printStackTrace();
                }
                //setProgressBar((double) counter++ / userAmount);
                double progress = (double) counter++ / userAmount;
                Platform.runLater(() -> progressBar.setProgress(progress));
                Platform.runLater(() -> progressLabel.setText("Progress: " + progress * 100 + "%"));
            }

            //Create a set of anime that will be removed
            for (String user : pap.getIgnoredUsers()) {
                try {
                    pap.fillSetOfIgnoredAnime(user);
                } catch (InterruptedException | IOException | SAXException | ParserConfigurationException e) {
                    e.printStackTrace();
                }
            }

            //Remove anime from map
            for (Integer animeID : pap.getIgnoredAnimeSet()) {
                pap.removeAnimeFromMergedAnimeListsMap(animeID);
            }
            pap.writeCSV();

            Platform.runLater(() -> maxCountLabel.setText("Max Count: " + pap.getMaxCount()));
            Platform.runLater(() -> averageScoreLabel.setText("Average Score: " + pap.getAverageScore()));
            Platform.runLater(() -> progressBar.setProgress(1.0));
            Platform.runLater(() -> progressLabel.setText("Done!"));
        }).start();
    }
}


src/parseAndProcess/ParserAndProcessor.java
package parseAndProcess;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.concurrent.TimeUnit;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.apache.commons.io.FileUtils;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

public class ParserAndProcessor {

    private int userCounter = 0;
    private boolean enteredUserToBeIgnored = false;
    private ArrayList<Integer> arrayListOfIgnoredEntries;
    private int sleepTime = 200;

    public static void main(String[] args) {
        ParserAndProcessor myInstance = new ParserAndProcessor();
        try {
            myInstance.createIgnoredAnimeList();
        } catch (IOException | SAXException | ParserConfigurationException e) {
            throw new RuntimeException(e);
        }
    }

    private void createIgnoredAnimeList() throws IOException, ParserConfigurationException, SAXException {
        Scanner sc = new Scanner(System.in);
        System.out.println("If you want the completed entries of a user to be excluded enter their name or don't and then press enter:");
        String ignoreCompletedUser = sc.nextLine();
        System.out.println("Enter amount of waiting in miliseconds between each user. Standard is 200. Increase if you are getting 429 errors (500ms should be enough), decrease if you think you are lucky.");
        sleepTime = sc.nextInt();
        if (sleepTime < 0) sleepTime = 200;
        sc.close();
        if (ignoreCompletedUser.length() > 0) {
            enteredUserToBeIgnored = true;
            String myAnimeListUserURL = "https://myanimelist.net/malappinfo.php?u=" + ignoreCompletedUser + "&status=all&type=anime";
            File xmlFile = new File("animelist", "tmp");
            URL url = new URL(myAnimeListUserURL);

            System.out.println("Processing entered User: " + ignoreCompletedUser);
            FileUtils.copyURLToFile(url, xmlFile);

            // Create the hashmap for this user
            ArrayList<Integer> animeList = new ArrayList<Integer>();

            // Parse the xml
            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(xmlFile);
            doc.getDocumentElement().normalize();
            NodeList nList = doc.getElementsByTagName("anime");

            // go through the parsed list and add it to the users list
            for (int i = 0; i < nList.getLength(); i++) {
                Node nNode = nList.item(i);

                // if its a valid anime node, create a row, extract the data and save it into the user
                // animelist
                if (nNode.getNodeType() == Node.ELEMENT_NODE) {
                    Element elem = (Element) nNode;
                    // only add anime if it is completed
                    if (Integer.parseInt(elem.getElementsByTagName("my_status").item(0).getTextContent()) == 2) {
                        Integer key = Integer.parseInt(elem.getElementsByTagName("series_animedb_id").item(0).getTextContent());
                        // put the anime into the row
                        animeList.add(key);
                        arrayListOfIgnoredEntries = animeList;
                    }
                }
            }
        }
        parseXML();
    }

    private void parseXML() throws IOException, ParserConfigurationException, SAXException {
        String readLine;
        String newReadLine = "";
        BufferedReader reader = new BufferedReader(new FileReader("malusers.txt"));
        ArrayList<HashMap<Integer, Row>> listOfMaps = new ArrayList<>();
        while ((readLine = reader.readLine()) != null) {
            int lastSlashPosition = readLine.lastIndexOf("/");
            if (lastSlashPosition != -1) {
                for (int i = lastSlashPosition + 1; i < readLine.length(); i++) {
                    newReadLine += readLine.charAt(i);
                }
                readLine = newReadLine;
                newReadLine = "";
            }
            HashMap<Integer, Row> aMap = getAndParseXmlForUser(readLine);
            listOfMaps.add(aMap);
        }

        HashMap<Integer, Row> combinedMap = listOfMaps.stream().reduce(new HashMap<>(), this::mergeMaps);

        reader.close();
        removeEntriesOfEnteredUser(arrayListOfIgnoredEntries, combinedMap);
    }

    private HashMap<Integer, Row> getAndParseXmlForUser(String user) throws IOException, ParserConfigurationException, SAXException {
        String myAnimeListUserURL = "https://myanimelist.net/malappinfo.php?u=" + user + "&status=all&type=anime";
        File xmlFile = new File("animelist", "tmp");
        URL url = new URL(myAnimeListUserURL);

        // sleep is necessary because the MAL API complains at too many requests
        // skip for first user
        try {
            TimeUnit.MILLISECONDS.sleep(sleepTime);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        userCounter++;
        System.out.println("Processing User #" + userCounter + ": " + user);
        FileUtils.copyURLToFile(url, xmlFile);

        // Create the hashmap for this user
        HashMap<Integer, Row> animeMap = new HashMap<>();

        // Parse the xml
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        Document doc = dBuilder.parse(xmlFile);
        doc.getDocumentElement().normalize();
        NodeList nList = doc.getElementsByTagName("anime");

        // go through the parsed list and add it to the users list
        for (int i = 0; i < nList.getLength(); i++) {
            Node nNode = nList.item(i);

            // if its a valid anime node, create a row, extract the data and save it into the user
            // animelist
            if (nNode.getNodeType() == Node.ELEMENT_NODE) {
                Element elem = (Element) nNode;
                // only add anime if it is not PTW
                if (Integer.parseInt(elem.getElementsByTagName("my_status").item(0).getTextContent()) != 6) {
                    Row row = new Row();
                    row.name = elem.getElementsByTagName("series_title").item(0).getTextContent();
                    row.count = 1;
                    // only add score and scorecount if score is not 0
                    if (Integer.parseInt(elem.getElementsByTagName("my_score").item(0).getTextContent()) == 0) {
                        row.score = 0;
                        row.scoreCount = 0;
                    } else {
                        row.score = Double.parseDouble(elem.getElementsByTagName("my_score").item(0).getTextContent());
                        row.scoreCount = 1;
                    }

                    Integer key = Integer.parseInt(elem.getElementsByTagName("series_animedb_id").item(0).getTextContent());
                    // put the anime into the row
                    animeMap.put(key, row);
                }
            }
        }
        return animeMap;
    }

    private HashMap<Integer, Row> mergeMaps(HashMap<Integer, Row> mapA, HashMap<Integer, Row> mapB) {
        HashMap<Integer, Row> newMap = new HashMap<>();
        newMap.putAll(mapA);

        for (Map.Entry<Integer, Row> entry : mapB.entrySet()) {
            if (newMap.containsKey(entry.getKey())) { // If the key is already in the map, we have a duplicate!

                // We need to update the existing entry

                // Only proceed if the user gave a rating
                if (entry.getValue().score != 0) {
                    if (newMap.get(entry.getKey()).score > 0) {
                        newMap.get(entry.getKey()).score = ((newMap.get(entry.getKey()).score * newMap.get(entry.getKey()).scoreCount) + entry.getValue().score) / (newMap.get(entry.getKey()).scoreCount + 1);
                        newMap.get(entry.getKey()).scoreCount += 1;
                    } else {
                        newMap.get(entry.getKey()).score = entry.getValue().score;
                        newMap.get(entry.getKey()).scoreCount = 1;
                    }
                }
                newMap.get(entry.getKey()).count += 1;
                // this should be it
            } else { // Key of map 2 is not in the new map, just add the non duplicate
                newMap.put(entry.getKey(), entry.getValue());
            }
        }
        return newMap;
    }

    private void removeEntriesOfEnteredUser(ArrayList<Integer> arrayListOfIgnoredEntries, HashMap<Integer, Row> combinedMap) {
        if (enteredUserToBeIgnored) {
            HashMap<Integer, Row> newMap = new HashMap<>();
            newMap.putAll(combinedMap);

            System.out.println("Removing entries");
            for (Integer elem : arrayListOfIgnoredEntries) {
                // if we find the same key in both maps, remove the entry
                if (newMap.containsKey(elem)) {
                    newMap.remove(elem);
                }
            }
            writeCSV(newMap);
        } else {
            writeCSV(combinedMap);
        }
}

    private void writeCSV(HashMap<Integer, Row> map) {
        double factorizedScore;
        double scoreCountNormalization;
        double weightedScore;
        final double FACTOR = 0.866;
        String link;
        int maxCount = 1;
        int titleAmountWithScore = 0;
        double totalScore = 0;

        for (Map.Entry<Integer, Row> entry : map.entrySet()) {
            if (entry.getValue().count > maxCount) {
                maxCount = entry.getValue().count;
            }
            if (entry.getValue().score > 0) {
                titleAmountWithScore++;
            }
            totalScore += entry.getValue().score;
        }

        System.out.println("Writing File");
        String result = "Name^Score^Count^Score Count^Weighted Score^Link" + "\n";

        for (Map.Entry<Integer, Row> entry : map.entrySet()) {
            link = "https://myanimelist.net/anime/" + entry.getKey();
            factorizedScore = (double) entry.getValue().score * FACTOR;
            scoreCountNormalization = (double) entry.getValue().count / maxCount;
            weightedScore = factorizedScore + scoreCountNormalization * (1.0 - FACTOR) * 10.0;
            result = result + entry.getValue() + "^" + weightedScore + "^" + link + "\n";
        }

        System.out.println("Max count: " + maxCount);
        System.out.println("Average rating: " + totalScore / titleAmountWithScore);

        try {
            FileUtils.writeStringToFile(new File("MergedLists.txt"), result, "UTF-8");
            System.out.println("Finished!");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
